Sources/Purchasing/StoreKit2/StoreKit2TransactionFetcher.swift

```diff
@@ -90,15 +115,147 @@ extension StoreKit.VerificationResult where SignedType == StoreKit.Transaction {
     var verifiedTransaction: StoreKit.Transaction? {
         switch self {
         case let .verified(transaction): return transaction
-        case .unverified: return nil
+        case let .unverified(transaction, error):
+            Logger.warn(Strings.storeKit.sk2_unverified_transaction(identifier: String(transaction.id), error))
+            return nil
         }
     }
 
-    fileprivate var verifiedStoreTransaction: StoreTransaction? {
+    var verifiedStoreTransaction: StoreTransaction? {
         switch self {
         case let .verified(transaction): return StoreTransaction(sk2Transaction: transaction,
                                                                  jwsRepresentation: self.jwsRepresentation)
-        case .unverified: return nil
+        case let .unverified(transaction, error):
+            Logger.warn(Strings.storeKit.sk2_unverified_transaction(identifier: String(transaction.id), error))
+            return nil
+        }
+    }
+
+}
+
+// MARK: - Private
+
+@available(iOS 16.0, macOS 13.0, tvOS 16.0, watchOS 9.0, *)
+extension StoreKit.VerificationResult where SignedType == StoreKit.AppTransaction {
+
+    var verifiedAppTransaction: SK2AppTransaction? {
+        switch self {
+        case let .verified(transaction): return .init(appTransaction: transaction)
+        case let .unverified(transaction, error):
+            Logger.warn(
+                Strings.storeKit.sk2_unverified_transaction(identifier: transaction.bundleID, error)
+            )
+            return nil
+        }
+    }
+
+}
+
+extension StoreKit2TransactionFetcher {
+
+    @available(iOS 15.0, tvOS 15.0, macOS 12.0, watchOS 8.0, *)
+    private var verifiedTransactions: [StoreTransaction] {
+        get async {
+            return await StoreKit.Transaction.all
+                .compactMap { $0.verifiedStoreTransaction }
+                .extractValues()
+        }
+    }
+
+    @available(iOS 15.0, tvOS 15.0, macOS 12.0, watchOS 8.0, *)
+    func verifiedTransactions(containing transaction: StoreTransactionType) async -> [StoreTransaction] {
+        // A newly generated transaction might not appear immediately in Transaction.all
+        var retries = 0
+        let sleepDuration: DispatchTimeInterval = .milliseconds(300)
+        let maximumRetries = 5
+
+        repeat {
+            retries += 1
+            let verifiedTransactions = await verifiedTransactions
+            if verifiedTransactions.contains(where: { $0.id == transaction.transactionIdentifier }) {
+                return verifiedTransactions
+            } else {
+                Logger.appleWarning(
+                    Strings.storeKit.sk2_receipt_missing_purchase(transactionId: transaction.transactionIdentifier)
+                )
+            }
+            try? await Task.sleep(nanoseconds: UInt64(sleepDuration.nanoseconds))
```

Comment:
Let's add a comment in this implementation to clarify that even though it's not efficient to introduce arbitrary delays, it's only applicable for SKTestSession / SK config files.